src/Data/Aeson/BetterErrors/Internal.hs

```diff
@@ -404,43 +425,49 @@ eachInObjectWithKey parseKey parseVal =
 -- | Lifts a function attempting to validate an arbitrary JSON value into a
 -- parser. You should only use this if absolutely necessary; the other
 -- functions in this module will generally give better error reporting.
-withValue :: (A.Value -> Either err a) -> Parse err a
-withValue f = liftParse (mapLeft CustomError . f)
+withValue :: (Functor m, Monad m) => (A.Value -> Either err a) -> ParseT err m a
+withValue f = liftParse (left CustomError . f)
 
-liftEither :: Either err a -> Parse err a
-liftEither = either (badSchema . CustomError) return
+liftEither :: (Functor m, Monad m) => Either err a -> ParseT err m a
+liftEither = withValue . const
 
-with :: Parse err a -> (a -> Either err b) -> Parse err b
-with g f = g >>= liftEither . f
+withM :: (Functor m, Monad m) => ParseT err m a -> (a -> m (Either err b)) -> ParseT err m b
+withM g f = g >>= lift . f >>= liftEither
 
-withText :: (Text -> Either err a) -> Parse err a
+with :: (Functor m, Monad m) => ParseT err m a -> (a -> Either err b) -> ParseT err m b
+with g f = withM g (return . f)
+
+withText :: (Functor m, Monad m) => (Text -> Either err a) -> ParseT err m a
 withText = with asText
 
-withString :: (String -> Either err a) -> Parse err a
+withString :: (Functor m, Monad m) => (String -> Either err a) -> ParseT err m a
 withString = with asString
 
-withScientific :: (Scientific -> Either err a) -> Parse err a
+withScientific :: (Functor m, Monad m) => (Scientific -> Either err a) -> ParseT err m a
 withScientific = with asScientific
 
-withIntegral :: Integral a => (a -> Either err b) -> Parse err b
+withIntegral :: (Functor m, Monad m, Integral a) => (a -> Either err b) -> ParseT err m b
 withIntegral = with asIntegral
 
-withRealFloat :: RealFloat a => (a -> Either err b) -> Parse err b
+withRealFloat :: (Functor m, Monad m, RealFloat a) => (a -> Either err b) -> ParseT err m b
 withRealFloat = with asRealFloat
 
-withBool :: (Bool -> Either err a) -> Parse err a
+withBool :: (Functor m, Monad m) => (Bool -> Either err a) -> ParseT err m a
 withBool = with asBool
 
 -- | Prefer to use functions like 'key' or 'eachInObject' to this one where
 -- possible, as they will generate better error messages.
-withObject :: (A.Object -> Either err a) -> Parse err a
+withObject :: (Functor m, Monad m) => (A.Object -> Either err a) -> ParseT err m a
 withObject = with asObject
 
 -- | Prefer to use functions like 'nth' or 'eachInArray' to this one where
 -- possible, as they will generate better error messages.
-withArray :: (A.Array -> Either err a) -> Parse err a
+withArray :: (Functor m, Monad m) => (A.Array -> Either err a) -> ParseT err m a
 withArray = with asArray
 
 -- | Throw a custom validation error.
-throwCustomError :: err -> Parse err a
+throwCustomError :: (Functor m, Monad m) => err -> ParseT err m a
 throwCustomError = liftEither . Left
+
+liftCustomT :: (Functor m, Monad m) => ExceptT err m a -> ParseT err m a
+liftCustomT f = lift (runExceptT f) >>= liftEither
```

Comment:
I noticed that this function isn't used anywhere else, and isn't exported by `Data.Aeson.BetterErrors`. I was planning that once I was happy enough with the API, I would hide the Internal module; the reason being I don't like things like the `MonadReader ParseReader` instance for `Parse` appearing in the docs.

Therefore, it shouldn't be necessary for "normal" use of this library to import the `Internal` module, I only exposed it for now as a sort of backdoor just in case. If this function is to become part of this library's API, I'd rather it were documented and re-exported by `Data.Aeson.BetterErrors`.